scRNA-seq-advanced/setup/integration-setup.Rmd

```diff
@@ -0,0 +1,402 @@
+  - This requires slightly more setup - you'd need to pass in a data frame of covariates, not just a batch vector
+- `harmony` does not calculate corrected expression values and only returns a matrix of corrected PCs
+- a bonus key way is that `harmony` is on CRAN but `batchelor` is on bioconductor.
+
+```{r}
+
+# harmony can either start with PCs or with normalized (uncorrected) expression
+# we already have PCs so let's just ask harmony to use those:
+harmony_result <- harmony::HarmonyMatrix(
+      # PCs
+      data_mat  = reducedDim(combined_sce, "PCA"), 
+      # batch information
+      meta_data = colData(combined_sce)$batch,
+      # don't run PCA since we gave PCs in `data_mat` argument
+      do_pca = FALSE
+    )
+
+# If we wanted to start with expression and have harmony calculate PCs for us:
+#harmony::HarmonyMatrix(
+#  data_mat  = logcounts(combined_sce), 
+#  meta_data = colData(combined_sce)$batch
+#)
+
+# The result is just a bunch of PCs!
+head(harmony_result)
+
+# Let's store this directly into our `combined_sce` object, making sure to name it well!
+reducedDim(combined_sce, "harmony_PCA") <- harmony_result
+
+# Calculate UMAP with the right name for PCs:
+# We do NOT want to overwrite the uncorrected UMAP! Make sure to name this differently with `name`
+combined_sce <- scater::runUMAP(combined_sce, dimred = "harmony_PCA", name = "harmony_UMAP")
+```
+
+#### The before/after reveal, again!
+
+
+```{r}
+# Before
+scater::plotReducedDim(combined_sce,
+                       dimred = "UMAP",
+                       colour_by = "batch",
+                       # Some styling to help us see the points:
+                       point_size = 0.5,
+                       point_alpha = 0.4) +
+  ggplot2::ggtitle("Before integration")
+
+
+# After
+scater::plotReducedDim(combined_sce,
+                       dimred = "harmony_UMAP",
+                       colour_by = "batch",
+                       # Some styling to help us see the points:
+                       point_size = 0.5,
+                       point_alpha = 0.4) +
+  ggplot2::ggtitle("After integration with harmony")
+
+```
+
+
+## Comparing back to celltypes
+
+In an ideal world, we'd have cell types to refer to in UMAPs as well; well-integrated datasets should show more clustering of cell types and less clustering of batches.
+Surprise, it's an ideal world!
+
+
+Before we look at cell types, it might be nice to have all our UMAPs in one place.
+Since we already have the unintegrated and harmony UMAPs in `combined_sce`, let's just go ahead and add `fastmnn` UMAPs into that SCE as well.
+We can use `combined_sce` going forward.
+```{r}
+reducedDim(combined_sce, "fastmnn_UMAP") <- reducedDim(integrated_sce, "UMAP")
+```
+
+
+Magic celltypes!
+Note that these subtypes only have the current libraries in them but this super quick to update with, for example, exercise libraries.
+Also note, we could go a bit rogue and start with SCEs that have the celltypes in `colData` already?
```

Comment:
Tagging in @allyhawkins to chat about the "prior prep" here since it involves the original Seurat objects. My hack setup script was parsing the RDS files that Ally already made by parsing Seurat!

Reply:
Yes, so the cell types come from the Seurat objects originally and we have a script in data integration that takes those cell types and adds them to the already filtered and normalized SCE objects. So I think there are two options: 

1. You start with those objects that have the cell types, are already filtered and already normalized and just read those in directly. The problem with this is that it's less transparent to trainees how the data was produced and how the data got from scpca to this notebook. 
2. The other option (which I think I favor in an effort to increase transparency), is to create a metadata table with the cell types as you have already done and read that into the set up script along with the `_filtered.rds` files straight from scpca. Then in the setup script you would do the filtering and normalization and dim reduction that you do here in addition to show how you added in the cell type column from the metadata table. This we would have to just make a metadata table first, which is the only part that wouldn't be transparent.

Reply:
I agree with this: starting with the table of  sample, cell_id, and cell type is a pretty good compromise in my mind. I don't think we can really share the Seurat objects, even if we wanted to... But we should have a clear acknowledgement about where the cell type info came from.